Simpler versions of `addCheck` and `addExecutable`

soenkehahn opened the pull request:
This is a somewhat drastic change, but the new overloads feel much better than the backtick version. I *think* for now we should still support the backtick version for cases where users want to splice in packages.

Also while I *think* this is a backwards compatible change, maybe we should still bump the ts lib version? WDYT?

jkarni commented:
Can you remind me why we can’t just use normal template literals (by writing custom toString methods that generate the nix string) and get rid of the tag function altogether?

alexdavid commented:
> Can you remind me why we can’t just use normal template literals (by writing custom toString methods that generate the nix string) and get rid of the tag function altogether?

The issue with this is that we can't do any escaping, a shell function that includes a quote, backslash, or `${...}` would be indistinguishable from an actual nix expression splatted in.

alexdavid commented on test/spec/CheckSpec.hs:
```diff
@@ -109,6 +109,27 @@ spec = do
+                  defaultEnvironment: garn.emptyEnvironment,
+                },
+                {},
+              ).addCheck("test-check")`echo error; false`;
```
Maybe this test should actually splat in a package to help document why we want to support this syntax still?

soenkehahn replied:
Yeah, good call! I did that in a0f09c5f. And also had to actually make it work. :)

alexdavid commented on ts/project.ts:
```diff
@@ -163,58 +179,80 @@ const proxyEnvironmentHelpers = (): ProjectHelpers => ({
-      ].join(" "),
-    );
-    return templateLiteralFn;
+    } else {
```
Nit, but maybe we remove this else? And the one in `addCheck` below?

soenkehahn replied:
Done in fd157e61.

alexdavid commented on ts/project.ts:
```diff
@@ -163,58 +179,80 @@ const proxyEnvironmentHelpers = (): ProjectHelpers => ({
+      const newExecutable = { [name]: this.shell(script) };
       return {
         ...this,
         ...newExecutable,
```
Interesting that you didn't run into the same `name` becoming `string` type degradation that i ran into with the extra function layer. Since this is the case, should we just inline it here?

```suggestion
        [name]: this.shell(script),
```

soenkehahn replied:
I think this might be because we updated the `deno` version in between?

Inlined in 4b19957c.

alexdavid reviewed (approved):
This looks great.

I was less concerned about `shell` and `check` using the template literals (in fact I think they're really  nice there), than the `addCheck` and `addExecutable` since it's kind of strange after a function call. But I don't feel strongly.

I think this PR should update the code on the website, examples, and readme before merging.

jkarni commented:
> The issue with this is that we can't do any escaping. A shell snippet that includes a quote, backslash, or ${...} would be indistinguishable from an actual nix expression splatted in.

Why though? We could have the toString be whatever we want - $$$THIS-IS-PATH-somepath$$$ - and then process the whole string before giving it to nix.

alexdavid commented:
Yeah, that's true, but it seems a little sad that we don't keep it structural in my opinion, especially since the user provides the structural data and we intentionally replace it with some magic string. Since we realized that splatting in packages is less of a common flow I think  having a specific tag template literal function that we can use feels like the right solution to me, but I don't feel too strongly if you want to keep them strings:

```typescript
something.addExecutable("run-cowsay", nix`${pkgs.cowsay}/bin/cowsay moo`)
```

soenkehahn commented:
> I think this PR should update the code on the website, examples, and readme before merging.

Good call, I updated the website and the readme in e673f18 (this one deserves scrutiny, since we don't typecheck this automatically anywhere) and updated the examples in 0459bef.

alexdavid reviewed (approved):
Pushed a nit review commit. Looks great!
